Implement concept of stream admin

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR is for implementing concept of stream admin,

- [x] Add a role field to subscription model. (merged)
- [x] Allow stream admins to update stream properties and to deactivate a stream.(merged)
- [x] Allow stream admin to remove others from stream.(merged)
- [x] Add `is_stream_admin` field to Subscription objects sent to clients.(merged)
- [x] Allow stream admin to change stream properties from UI.
- [x] Prevent unsubscribing of last stream admin.
- [x] Change role from stream admin to role if stream admin unsubscribes.
- [x] Make creator of stream a stream admin.
- [x] Add API endpoint to change the role of user in stream.
- [x] Add `stream_admins` field to Subscription objects passed to clients.
- [ ] Add indicator for stream admin subscribers list.
- [ ] UI for changing subscription role.

<!-- How have you tested? -->


<!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

clarammdantas commented on zerver/lib/actions.py:
```diff
@@ -2734,7 +2734,8 @@ def bulk_add_subscriptions(streams: Iterable[Stream],
                            from_stream_creation: bool=False,
-                           acting_user: Optional[UserProfile]=None) -> SubT:
+                           acting_user: Optional[UserProfile]=None,
+                           stream_admin_map: Optional[Dict[int, List[Stream]]]=None) -> SubT:
```
Wouldn't be better if instead of `List[Stream]` we had `Set[Stream]`? I think it makes sense to have a list of unique streams where that user is an admin, and for that, a set would be more appropriate. Also, making the check `stream in stream_admin_map[user_profile.id]` below would be faster.

timabbott replied:
`Set` doesn't work well with Django model objects, so we shouldn't use that data structure.  A `Dict` mapping a stream_id to the object could be used if we preferred that for some reason, but I'm not sure why we would.

sahil839 commented:
@zulipbot remove "has conflicts"

sahil839 commented on static/js/stream_data.js:
```diff
@@ -478,17 +478,19 @@ exports.receives_notifications = function (stream_id, notification_name) {
 exports.update_calculated_fields = function (sub) {
     sub.is_admin = page_params.is_admin;
+    sub.can_perform_admin_actions = page_params.is_admin || sub.is_stream_admin;
     // Admin can change any stream's name & description either stream is public or
```
`sub.is_admin` does not feel correct because it may be interpreted as stream admin or realm admin both. 
I have not changed/removed it yet because it is used to check whether the user is allowed to delete or messages in topic or move a topic and wanted to have a discussion on whether we would allow stream admins to perform these tasks. We can change this after we have decided or we can just replace the places where we use `sub.is_admin` by `page_params.is_admin`

timabbott replied:
Yes, agreed.  We might want to head off this sort of confusion on the other end by renaming `sub.is_admin` to `sub.is_realm_admin` everywhere in a preparatory refactoring commit.  (Or maybe we can just remove that field altogether?  I'm not sure it needs to be copied onto stream objects)?

`sub.can_perform_admin_actions` should be `sub.can_administer_stream`, I think.

So maybe the refactoring is just a prep commit that does

s/sub.is_admin/sub.can_administer_stream/, and then this commit just adds `sub.is_stream_admin` and changes how `sub.can_administer_stream` is calculated?

sahil839 replied:
I think we can just replace sub.is_admin with page_params.is_admin. There are only few places where we use this.

timabbott replied:
Remember to check `git grep is_admin static/templates/` -- we might pass a `sub` to a template in some places.

But I think my proposal achieves that same goal while having a nice valid `sub.can_administer_stream` that we'll find useful.

timabbott replied:
Looks like `static/templates/subscription_settings.hbs` has it -- so that's why it's there.  In any case I think we're agreed we should just remove `sub.is_admin` in favor of something better.

sahil839 replied:
`is_admin` in `static/templates/subscription_settings.hbs` will be replcaed by `can_administer_stream` as we would allow stream admin to deactivate the stream.
I have checked and the only place we use is_admin is in `static/templates/topic_sidebar_actions` and in `stream_popover.js` to render the template mentioned above.

sahil839 commented:
This has not been decided and needs to be discussed.

pragatiagrawal31 commented on static/templates/subscription_settings.hbs:
```diff
@@ -9,15 +9,11 @@
-                {{#if is_admin}}
-                <button class="button small rounded btn-danger deactivate" type="button" name="delete_button" title="{{t 'Delete stream'}}"> <i class="fa fa-trash-o" aria-hidden="true"></i></button>
-                {{/if}}
+                <button class="button small rounded btn-danger deactivate" type="button" name="delete_button" title="{{t 'Delete stream'}}" {{#unless can_delete_stream}}style="display: none"{{/unless}}> <i class="fa fa-trash-o" aria-hidden="true"></i></button>
```
I recall I had a review in one of my PRs or maybe somewhere that we refrain from using inline CSS and use it when there is no other option (we use it in emails for obvious reasons). Can we use `hide` class here?

sahil839 replied:
I will change this.

pragatiagrawal31 commented on static/js/stream_ui_updates.js:
```diff
@@ -93,6 +93,22 @@ exports.update_regular_sub_settings = function (sub) {
+    }
+    const $settings = $(".subscription_settings[data-stream-id='" + sub.stream_id + "']");
+    if (sub.can_perform_admin_actions) {
+        $settings.find(".stream-name .editable").show();
+        $settings.find(".stream-description .editable").show();
+        $settings.find(".deactivate").show();
```
I will try to represent all three elements simultaneously using a class (a new or common one).
(Although this code is likely to change by one of my PR in the future.)

sahil839 replied:
We can add a new class here like stream-settings-header-controls.

sahil839 commented:
@timabbott this is ready for review.

timabbott commented:
@sahil839 thanks for your patience with this getting cut from the 3.0 release plan.  Looking at it now.

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -10,6 +10,14 @@ below features are supported.
+  `is_stream_admin` field to Subscription objects.
+
+* [`GET /events`](/api/get-events): Added `is_stream_admin` field to
+  Subscription objects sent in `subscriptions` event with `op=add`.
```
These should go in the Zulip 4.0 block, once you've rebased this.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2826,8 +2827,12 @@ def bulk_add_subscriptions(streams: Iterable[Stream],
+        role = Subscription.ROLE_MEMBER
+        if stream_admin_map is not None:
+            if user_profile.id in stream_admin_map and stream in stream_admin_map[user_profile.id]:
+                role = Subscription.ROLE_STREAM_ADMINISTRATOR
```
I think I'd prefer to have the commit that makes it possible to create stream administrators sit later in the branch, after we've polished the changes that implement the basic logic for what they can do.  Can you move this commit a few steps later?  

Does this change result in `populate_db` creating any streams with a stream administrator?

sahil839 replied:
I have moved the commit. 
Currently I have not changed the code in populate_db to make stream administrator, will do this.

timabbott commented on zerver/lib/streams.py:
```diff
@@ -199,22 +199,33 @@ def check_for_exactly_one_stream_arg(stream_id: Optional[int], stream: Optional[
+        return
+
+    if sub is not None and sub.is_stream_admin:
+        return
```
I think we need to check if the user is subscribed, and return "Invalid stream id" if it's a private stream the user isn't subscribed to; otherwise this endpoint supports checking if a private stream exists.

timabbott commented on zerver/lib/streams.py:
```diff
@@ -436,6 +448,19 @@ def list_to_streams(streams_raw: Iterable[Mapping[str, Any]],
+            sub = None
+            if stream.id in sub_dict_by_stream_ids:
+                sub = sub_dict_by_stream_ids[stream.id]
+            check_stream_access_for_delete_or_update(user_profile, stream, sub)
```
Well done using this structure to avoid database queries in a loop here.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -374,48 +375,86 @@ class StreamAdminTest(ZulipTestCase):
+        }
+        stream = self.subscribe(user_profile, 'private_stream_2')
+        result = self.client_patch(f"/json/streams/{stream.id}", params)
+        self.assert_json_error(result, "This action requires administrative rights.")
```
Let's have this error message be formatted like the one for the realm administrators; can you make an exception class like  `OrganizationAdministratorRequired` called `StreamAdministratorRequired` with message "Must be an organization or stream administrator" for this?

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -374,48 +375,86 @@ class StreamAdminTest(ZulipTestCase):
+        }
+        stream = self.subscribe(user_profile, 'public_stream_2')
+        result = self.client_patch(f"/json/streams/{stream.id}", params)
+        self.assert_json_error(result, "This action requires administrative rights.")
```
Let's add a chekc here that `stream.invite_only` is False (Etc.) -- that the change didn't happen in addition to it throwing the error.  Same with the last test.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -715,15 +768,31 @@ def test_rename_stream(self) -> None:
+        self.assert_json_success(result)
+
+        stream_rename_exists = get_stream('stream_rename', realm)
+        self.assertTrue(stream_rename_exists)
```
I don't understand this test -- we don't seem to inspect `events` after we're done?

timabbott replied:
We should at least assert the number of events is as expected.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -1220,78 +1323,113 @@ def test_cant_remove_others_from_stream(self) -> None:
-            target_users=[self.example_user('cordelia')], target_users_subbed=True,
+            query_count=6, is_realm_admin=False, is_stream_admin=False, is_subbed=True,
+            invite_only=False, target_users=[self.example_user('cordelia')], target_users_subbed=True,
             using_legacy_emails=True)
```
Do you understand why this adds 3 database queries?  I'd have expected just 1 extra query.

sahil839 replied:
I did not notice it earlier. But I am also not able to understand why 3 queries are being added.

timabbott replied:
Well, the best way to check is to print out the queries before-and-after and see.

timabbott replied:
The query count bug was fixed by this:

```
-        subs = Subscription.objects.filter(user_profile=user_profile,
-                                           recipient__type=Recipient.STREAM,
-                                           recipient__type_id__in=existing_stream_ids,
-                                           active=True)
+        subs = Subscription.objects.select_related("recipient").filter(
+            user_profile=user_profile,
+            recipient__type=Recipient.STREAM,
+            recipient__type_id__in=existing_stream_ids,
+            active=True)
```

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -1220,78 +1323,113 @@ def test_cant_remove_others_from_stream(self) -> None:
+
+    def test_stream_admin_remove_others_from_public_stream(self) -> None:
+        """
+        If you're a stream admin, you can remove people from public streams.
```
Reword this; should say "You can remove people from public streams you're a stream administrator for."

And same for the others here.  Critically, the behavior is only for the streams you administer, and we should make sure these comments don't mislead on what we're testing.

timabbott commented:
This is looking good!  I think the first commit may by mergable; I'll do so if it is.  Once that's done:
* I posted I think a full set of comments on the commit I'd like to merge next, namely the one to ensure stream administrators (in that commit an empty set except in tests that create one) have the administrative access to the stream they have.  We can merge that into master as soon as it's ready without other changes.
* Next, we'll want to integrate "streams: Add is_stream_admin field to sub objects passed to clients.".  I think we should change that to pass `role` in the API rather than `is_stream_admin`; I hope we've learned our lesson from the user-level role field.  And you'll need to bump API_FEATURE_LEVEL; may be more or less a rewrite of that commit.  

Let me know when you have those commits ready for a next review.  I think we'll be able to merge the bulk of this PR this week, though we may not be able to finish the "Stream administrators" feature until we resolve questions like "Must a stream have an administrator?" and figure out how to do UI for changing roles in streams.

sahil839 commented:
@timabbott Updated the PR with following changes-
- Added check for unsubscribed user trying to access private stream. I was confused where to add the tests for this becuase we allow realm admins to update unsubscribed private streams anyway and also currently we have restriction that stream admin is always the user who is subscribed for public as well as private streams. So I have added the tests for this as a separate test as `test_non_admins_cannot_access_unsub_private_stream` for now.
- Added StreamAdministratorRequired exception class.
- And some other changes in tests that you suggested.

Regarding the `is_stream_admin` field passed to clients, we currently pass boolean fields only ('is_admin`, `is_owner`) to clients, while clients send `role` to API for changing it. So, wanted to confirm whether we want to pass `is_stream_admin` or `role` to clients.

timabbott commented:
I merged the first two commits as the series ending with ca1a8ac78f5863e5fd1241b0259871b5a2849607.  In theory, that's most of the complexity on the backend.  Can you proceed with making the API send `role`?  That should be the next commit for this.

sahil839 commented:
@timabbott I have updated the API commit to send role.

sahil839 commented:
Currently, we do not change role of stream admin to member of unsubscribing in the original Subscription object, we only change the role to member in stream_dict sent to API. 
I will update this to change the role itself on unsubscribing, after seeing how other products handle this.

timabbott commented:
Merged the feature as f046c9c58a24a8ad5cae8c79a71100648e29ec05 after a bit of polish on the API documentation and changelog.

timabbott commented on frontend_tests/node_tests/stream_data.js:
```diff
@@ -424,23 +424,43 @@ run_test("admin_options", () => {
     // non-admins can't do anything
     global.page_params.is_admin = false;
     let sub = make_sub();
+    sub.role = stream_data.sub_role_values.member.code;
```
"frontend" is mispelled in this commit's title :)

timabbott commented on static/templates/stream_member_list_entry.hbs:
```diff
@@ -3,7 +3,7 @@
     {{#if show_email}}
```
We'll want to put some indication of which users are stream administrators in this table.

sahil839 commented:
@timabbott I have made the following changes-
- Added commit for preventing last stream admin from unsubscribing.
- Added commit for changing role to member when stream admin unsubscribes.
- Renamed sub.is_admin to sub.is_stream_admin
- Made the suggested changes in frontend commit.

Next, I will be working on adding endpoint for changing role and adding an indication for stream admin in subscribers list.

Regarding showing stream admin in subscribers list, I am thinking of two options- adding a separate role column or adding stream admin besides the name and nothing for members like `Iago (stream administrator)`. Any suggestions on this or some other ideas?

sahil839 commented:
@zulipbot remove "has conflicts"

sahil839 commented:
@timabbott This PR is ready for the next review. Updated the PR with following -
- Added the commit for preventing last stream admin to unsubscribe. Currently, this restriction is for all the streams. We can change it if we want to keep it only for private streams.
- Changing role from stream admin to member on unsubscribing.
- Added API endpoint for changing role.

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -3066,6 +3079,16 @@ def get_non_subscribed_tups() -> List[Tuple[UserProfile, Stream]]:
+                     value=Subscription.ROLE_MEMBER,
+                     stream_id=stream.id,
+                     name=stream.name)
+        send_event(user_profile.realm, event, can_access_stream_user_ids(stream))
```
@timabbott At this place database queries for `can_access_stream_user_ids` are performed in a loop. I am not able to figure out how to avoid this queries in loop.

sahil839 commented:
@timabbott Rebased the PR. Please  review.

timabbott commented:
I merged a couple more prep commits as the series ending with ab90abfa4c316197b85d0d55b4b63d0af0a1e2e6.  In the future, it'd be great if you can put independent refactors like those at the start of the branch so they're easy to find :).

sahil839 commented:
@timabbott rebased the PR with the following changes-
- Moved a prep commit to extract `check_subscription_property_update` to the top.
- Squashed the tweak commit you added (also fixed the typo in that commit).

sahil839 commented:
@timabbott Added a commit to allow stream admins to post to stream irrespective of stream_post_policy. Please review this PR.

timabbott commented:
Can you rebase?  I'll also merge what I can now.

sahil839 commented:
@timabbott rebased.

sahil839 commented:
@timabbott I have rebased the PR. Please review.

sahil839 commented:
@zulipbot remove "has conflicts"

andersk commented:
From the [discussion on CZO](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/openapi.20error/near/994668), this was blocked on p1c2u/openapi-core#265 somehow, so FYI we pulled in the fix for that with #17266.

sahil839 commented:
> From the [discussion on CZO](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/openapi.20error/near/994668), this was blocked on [p1c2u/openapi-core#265](https://github.com/p1c2u/openapi-core/issues/265) somehow, so FYI we pulled in the fix for that with #17266.

This PR was not blocked for that as I had already made a workaround commit for changing the order of those endpoints which fixed that error and that commit is already merged to master.

sumanthvrao commented on zerver/lib/actions.py:
```diff
@@ -1181,6 +1181,12 @@ def do_deactivate_stream(
     affected_user_ids = can_access_stream_user_ids(stream)
 
+    Subscription.objects.filter(
+        recipient__type=Recipient.STREAM,
```
Since this part handles the case of changing admin roles for deactivated streams, it doesn't exactly fall under "Unsubscribing stream admin" case as the commit indicates. Maybe want to update the title / move this to another commit?

sahil839 replied:
I have added this because deactivating a stream first unsubscribes all the members using this line `get_active_subscriptions_for_stream_id(stream.id).update(active=False)`. Maybe I can specify this in commit message description.

sumanthvrao commented on zerver/tests/test_subs.py:
```diff
@@ -2771,6 +2809,10 @@ def test_basic_add_delete(self) -> None:
         streams = self.get_streams(user)
         self.assertTrue("my_test_stream_1" in streams)
 
+        stream = get_stream("my_test_stream_1", user.realm)
```
This seemed to be a little confusing. If a user creates a stream (and becomes the stream admin for it in the process), then doesn't their role entitle them to delete the stream? Converting the admin to a member, before deleting the stream seems a little odd.

sahil839 replied:
Actually the url used in this test is not for deleting/deactivating the stream, but instead of unsubscribing the user, which in turn deletes the stream as there are no subscribers. So I am changing the role to member, as we do not allow last stream admin to unsubscribe. Maybe the confusion was there due to comment in that test, we can correct that comments.

sahil839 replied:
On second thought, we can remove the restriction of removing of not allowing the last admin to unsubscribe if the stream admin is only subscriber in the stream. What are your thoughts on this?

sumanthvrao replied:
Yup exactly. We should just allow the last member (who would be an admin as per our previous conditions) to delete the stream (i.e unsubscribe) without changing their role to a member. Want to bring this up on czo? :)

sahil839 replied:
Started the converstaion https://chat.zulip.org/#narrow/stream/3-backend/topic/stream.20administrators/near/1122637 on czo.

timabbott commented:
For `get_stream_admin_dict_for_streams`, I think we'll want to add a database index including the `role` property; the most natural would be an index on `recipient, role, active`.  We should test what query plan that generates for `get_stream_admin_dict_for_streams`; it's possible we'll want to change the query to be interacting with `recipient_ids` instead.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -3562,6 +3563,20 @@ def get_non_subscribed_subs() -> List[Tuple[UserProfile, Stream]]:
+
+    for sub in subs_to_deactivate:
+        user_profile_id = sub.user.id
+        stream_id = sub.stream.id
```
This should be `for sub_info in ...` ; we just `sub ` as a local variable only for Subscription objects, not SubInfo objects.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -3562,6 +3563,20 @@ def get_non_subscribed_subs() -> List[Tuple[UserProfile, Stream]]:
+        if user_profile_id in stream_admin_dict[stream_id]:
+            if stream_admin_count_dict[stream_id] == 1:
+                raise JsonableError(_("Cannot remove all stream administrators."))
+            stream_admin_count_dict[stream_id] -= 1
```
Let's move this block to a function `check_last_stream_admin_cannot_unsubscribe`.  I think we could also implement that function a bit more simply as e.g.:
```
stream_admin_dict = get_stream_admin_dict_for_streams(stream_ids) # Returns a Dict[id, Set[id]]
for sub in subs_to_deactivate:
    stream_admin_dict[stream_id].remove(sub.user_profile_id)

for (stream_id, remaining_admins) in stream_admin_dict.items():
    if len(remaining_admins) == 0:
        raise ...
```

timabbott commented on zerver/lib/streams.py:
```diff
@@ -642,3 +642,20 @@ def get_stream_by_narrow_operand_access_unchecked(operand: Union[str, int], real
     return get_stream_by_id_in_realm(operand, realm)
+
+
+def get_stream_admin_dict_for_streams(stream_ids: List[int]) -> Dict[int, List[int]]:
```
It might be cleaner for this to accept a list of `recipient_id`s just to make the query below cleaner, in combination with the more natural `id, role, active` database index we'd want to add.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -1605,10 +1605,10 @@ def test_stream_admin_remove_multiple_users_from_stream(self) -> None:
         ]
         result = self.attempt_unsubscribe_of_principal(
-            query_count=31,
+            query_count=33,
```
Do you understand why this sometimes adds two queries?

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1181,6 +1181,12 @@ def do_deactivate_stream(
+        recipient__type=Recipient.STREAM,
+        recipient__type_id=stream.id,
+        role=Subscription.ROLE_STREAM_ADMINISTRATOR,
+    ).update(role=Subscription.ROLE_MEMBER)
```
Hmm, ideally we'd want this to still record something in RealmAuditLog about the change in administrator status, e.g. by calling the `do_...` function for changing role, rather than a raw database query.

We haven't settled plans for reactivating deactivated streams, but if/when we do, we may want a record of who the administrators had been.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -3569,27 +3575,54 @@ def get_non_subscribed_subs() -> List[Tuple[UserProfile, Stream]]:
+                extra_data=orjson.dumps(
+                    {
+                        RealmAuditLog.OLD_VALUE: Subscription.ROLE_STREAM_ADMINISTRATOR,
+                        RealmAuditLog.NEW_VALUE: Subscription.ROLE_MEMBER,
```
Can we add a "reason: "unsubscribe""` field in this block?  I imagine that detail may be important for future investigation.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -3614,6 +3647,18 @@ def get_non_subscribed_subs() -> List[Tuple[UserProfile, Stream]]:
+            value=Subscription.ROLE_MEMBER,
+            stream_id=sub.stream.id,
+            name=sub.stream.name,
+        )
```
This should definitely call a `notify_change_stream_role` function that will be shared with the code path for changing stream-level roles; otherwise we're going to end up with bugs where we update one of the copies of this event format and not others.

Also, we should not send email/stream_name fields (we want to work only with `id`s instead) and actually send a list of user_ids and stream_ids, so we can use a single bulk event where convenient, even if we're doing a loop for simplicity of implementation, so that we don't need to change the protocol to later optimize this.  (We had real performance issues with this type of code path in the past, though with more common operations).

sahil839 replied:
@timabbott I agree with sending only ids in event, should we also change the event dict sent from `do_change_subscription_property` to only send id and not user email or stream name. I am suggesting this because event sent from `do_change_subscription_role` and `do_change_subscription_property` have same event `type=subscription` and `op=update` and the openapi schema is not validating in tests because both functions send different details in event objects. There is another way by sending the different event type for role case, but I think we would eventually prefer only ids and not stream name and user email.

timabbott commented on templates/zerver/api/changelog.md:
```diff
@@ -10,6 +10,15 @@ below features are supported.
+  admins of the stream.
+
+* [`GET /events`](/api/get-events): Added `stream_admins` field to Subscription
+  objects sent in `subscriptions` events.
```
Let's move this commit earlier in the branch, before any code that makes it possible to create stream administrators (including when one creates a new stream); that's generally the correct merge order.

timabbott commented on static/templates/subscription_settings.hbs:
```diff
@@ -29,18 +25,16 @@
                   message_retention_text=../message_retention_text}}
             </div>
-            <a class="change-stream-privacy" {{#unless can_change_stream_permissions}}style="display: none;"{{/unless}}>[{{t "Change" }}]</a>
+            <a class="change-stream-privacy {{#unless can_change_stream_permissions}}hide{{/unless}}">[{{t "Change" }}]</a>
```
A lot of this refactoring looks like it could be done in prep commits that just move us to declarative variables that initially still equal `is_realm_admin`; I think that'd be well worth doing to reduce the complexity of this commit.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2733,9 +2733,26 @@ def check_message(
+            except Subscription.DoesNotExist:
+                sub = None
+
+        # This will raise JsonableError if there are problems.
```
Hmm, what's the thinking around this logic being here rather than in `access_stream_for_send_message`?  It doesn't look like we end up passing the `sub` object.

timabbott reviewed (commented):
Thanks for all your work on this @sahil839!  I posted a batch of comments after which I hope we'll be able to merge a substantial prefix of this PR :).

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -3019,19 +3040,55 @@ def get_non_subscribed_tups() -> List[Tuple[UserProfile, Stream]]:
+            if stream_admin_count_dict[stream.id] == 1:
+                raise JsonableError(_("Cannot remove all stream administators."))
+            stream_admin_count_dict[stream.id] -= 1
+            removed_stream_admin_dict.append((sub.user_profile, stream))
```
I did it this way as we in other places also, like where one of the user_id is incorrect, we do not proceed and raise the error.
Also, here there can be multiple streams and multiple users, so I am checking for each stream whether all the admins of that streams are removed. Do you mean I should instead have a loop on stream and then check wheter the set of users of for that stream contains the set of admins or not?

Also, I cannot find any commit here for squashing.

sahil839 commented:
> @timabbott this is ready for review. Have done the following updates -
> 
> * Added the index (recipient_id, role, active)
> * Changed `get_stream_admin_dict` to accept `recipient_ids`. So this also led to change in the return type of this and we send dict of the form `{'recipient_id': list of stream admins}` now, we can use dict of form `stream_id: list of stream admins` now also but will involve and extra query.
> * I cannot understand the point here of incorrect logic [here](https://github.com/zulip/zulip/pull/15334#discussion_r488386823), it handles the case of streams with no stream admins also.
> * Added `user_profile__is_active=True` in `get_stream_admin_dict` query.
> * Extracted the `check_last_admin_cannot_unsubscribe` function.
> * Added `reason: unsubscribe` in the realmaudit log entries. Do we also need to add `reason: deactivate` for deactvating stream cases. Also the current models for`RealmAuditLog` has constant values for OLD_VALUE, etc. fields in extra data. Do we also want to add `reason` as a constant in the RealmAuditLog table and then use it or using string "reason" is fine.
> * Added RealmAuditLog entries also where it was required as mentioned by you.
> * Reordered the commits and also added a prep commit for frontend change.
> * Moved the logic of checking role in `access_send_stream_message` (cannot remember the reason why I did it in `check_message` before)
> 
> One change is left which is to allow the stream admin to unsubscribe if there is only one subscriber of the stream which is stream admin.

Last push was to just move the prep commit for refactoring at the top.

sahil839 commented:
> > @timabbott this is ready for review. Have done the following updates -
> > 
> > * Added the index (recipient_id, role, active)
> > * Changed `get_stream_admin_dict` to accept `recipient_ids`. So this also led to change in the return type of this and we send dict of the form `{'recipient_id': list of stream admins}` now, we can use dict of form `stream_id: list of stream admins` now also but will involve and extra query.
> > * I cannot understand the point here of incorrect logic [here](https://github.com/zulip/zulip/pull/15334#discussion_r488386823), it handles the case of streams with no stream admins also.
> > * Added `user_profile__is_active=True` in `get_stream_admin_dict` query.
> > * Extracted the `check_last_admin_cannot_unsubscribe` function.
> > * Added `reason: unsubscribe` in the realmaudit log entries. Do we also need to add `reason: deactivate` for deactvating stream cases. Also the current models for`RealmAuditLog` has constant values for OLD_VALUE, etc. fields in extra data. Do we also want to add `reason` as a constant in the RealmAuditLog table and then use it or using string "reason" is fine.
> > * Added RealmAuditLog entries also where it was required as mentioned by you.
> > * Reordered the commits and also added a prep commit for frontend change.
> > * Moved the logic of checking role in `access_send_stream_message` (cannot remember the reason why I did it in `check_message` before)
> > 
> > One change is left which is to allow the stream admin to unsubscribe if there is only one subscriber of the stream which is stream admin.
> 
> Last push was to just move the prep commit for refactoring at the top.

This push was for fixing a variable name, now this is ready for review. WIll add the code to allow stream admins to unsubscribe if all the users are being unsubscribed in a couple of days.

sahil839 commented:
@timabbott this is ready for review. Have done the following updates -

- Added the index (recipient_id, role, active)
- Changed get_stream_admin_dict to accept recipient_ids. So this also led to change in the return type of this and we send dict of the form {'recipient_id': list of stream admins} now, we can use dict of form stream_id: list of stream admins now also but will involve and extra query.
- I cannot understand the point here of incorrect logic here, it handles the case of streams with no stream admins also.
- Added user_profile__is_active=True in get_stream_admin_dict query.
- Extracted the check_last_admin_cannot_unsubscribe function.
- Added reason: unsubscribe in the realmaudit log entries. Do we also need to add reason: deactivate for deactvating stream cases. Also the current models forRealmAuditLog has constant values for OLD_VALUE, etc. fields in extra data. Do we also want to add reason as a constant in the RealmAuditLog table and then use it or using string "reason" is fine.
- Added RealmAuditLog entries also where it was required as mentioned by you.
- Reordered the commits and also added a prep commit for frontend change.
- Moved the logic of checking role in access_send_stream_message (cannot remember the reason why I did it in check_message before)
One change is left which is to allow the stream admin to unsubscribe if there is only one subscriber of the stream which is stream admin.
Will try to get it done ASAP, but would be nice to get a review here till then.  (Posting the list again here as faced some issue in GitHub mobile app which edits the comment on selecting Quote and reply option)

timabbott commented:
I think we can probably close this, since our user groups plans in https://github.com/zulip/zulip/issues/19525 replace it. 

There's probably cleanup to do, in the form of deleting `Subscription.role`?

timabbott commented:
Close, since we've now removed `subscription.role`.
